xcodeproj/internal/xcodeproj_rule.bzl

```diff
@@ -589,21 +603,34 @@ targets.
 
     non_mergable_targets = {}
     non_terminal_dests = {}
-    for src, dests in target_merges.items():
-        src_target = focused_targets[src]
-
-        for dest in dests:
-            dest_target = focused_targets[dest]
+    for dest, srcs in target_merge_dests.items():
+        dest_target = focused_targets[dest]
 
-            if dest_target.product.type not in _TERMINAL_PRODUCT_TYPES:
+        if dest_target.product.type not in _TERMINAL_PRODUCT_TYPES:
+            for src in srcs:
                 non_terminal_dests.setdefault(src, []).append(dest)
 
-            for library in xcode_targets.get_top_level_static_libraries(
-                dest_target,
-            ):
+        src_targets = [focused_targets[src] for src in srcs]
+        src_labels = [src_target.label for src_target in src_targets]
+
+        for library in xcode_targets.get_top_level_static_libraries(
+            dest_target,
+        ):
+            for src_target in src_targets:
+                ###############################
+                # FIXME
+                #
+                # Make this section more terse since it's doing duplicated work/might be wrong.
+                ###############################
+
+                # If direct owner, ignore since this is expected for a given library.
                 if library.owner == src_target.label:
                     continue
 
+                # If all src_targets have same dest_target, consider good as well.
+                if library.owner in src_labels:
+                    continue
```

Comment:
Transfer from https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2058. This check should ideally be "all srcs for this dest are going to this-and-only-this one dest", IMO.